### PR TITLE
Add save! and save methods

### PR DESCRIPTION
**Pull request by derigible**
Break out the save method to match more closely with
Ruby style of <method>! to raise an error while
<method> fails silently. Implemented to work like
ActiveRecord's save where an unsuccessful save will
mark the record as invalid.

### CODE REVIEW

File: lib/aws-record/record/item_operations.rb [hunk 1]
```diff
@@ -27,13 +28,29 @@ def self.included(sub_class)
...
+        save!
+      rescue Errors::KeyMissing => e
+        errors << e.message
+      end
+
+      # Is the record a valid record. True if #save was successful, false if
+      # Aws::Record::Errors::KeyMissing when using #save.
+      #
+      def valid?
+        errors.empty?
```
Comment (awood45) [inline lib/aws-record/record/item_operations.rb]: To have this functionality, you're going to need to be able to flush the `errors` array on a successful save. Right now, an invalid write which is then corrected will still show invalid. Add a test for that case too (invalid write, correction, valid write, `#valid?` shows as true after second attempt).

**Comment by awood45** [comment]
I'm good with the concept, but +1 to @tierra comment on changing the return case. Also should test that the `errors` array is emptied by a successful write.

Finally, this will also break if save hits exceptions other than `Errors::KeyError`. This should definitely capture anything in the `Aws::Record::Errors` namespace, and we possibly may want to capture and record `Aws::DynamoDB::Errors::ServiceError` (+children) and possibly `Seahorse::Client::NetworkingError`, which can be surfaced in some cases (such as networking issues from your host to DynamoDB).

**Comment by awood45** [comment]
Thanks for that. For now, let's capture all `Aws::Record::Errors` errors and ancestors, and I'll look into adding more.

**Comment by derigible** [comment]
I couldn't figure out a way to set up a situation where some of the other errors (like NameCollision) would be able to propogate to the save. It seems like these errors would stop the workflow before save was ever given a chance to be called. But if there is a scenario I am missing i will add it to the specs.

**Comment by awood45** [comment]
You're right to notice that there aren't really other cases likely to occur in the wild right now. However, I'm looking at having `save` respect dirty attributes and call `#update_item` when possible. Other features like optimistic locking could raise other errors that are not yet defined. An error hierarchy as used in your last commit is appropriate, though I'll likely be modifying it before GA.

**Comment by awood45** [comment]
That said, this looks good now. Thanks!

**Pull request merged**